### PR TITLE
Delete operation vault

### PR DESCRIPTION
**Pull request by MaazMS**
create Delete method  vault 

* create and delete logs 
```bash
INFO	controllers.Workshop	Adding Finalizer for the Workshop	{"workshop": "workshop-infra/cloud-native-workshop"}
INFO[0022] Created redis Secret                          source="portal_reconciler.go:51"
INFO[0023] Created redis Persistent Volume Claim         source="portal_reconciler.go:58"
INFO[0023] Created redis Deployment                      source="portal_reconciler.go:66"
INFO[0023] Created portal Deployment                     source="portal_reconciler.go:109"
INFO[0024] Created portal Route                          source="portal_reconciler.go:139"
INFO[0025] Created gitea Project                         source="gitea_reconciler.go:73"
INFO[0026] Created gitea-operator Service Account        source="gitea_reconciler.go:89"
INFO[0026] Created gitea-operator Operator               source="gitea_reconciler.go:114"
INFO[0026] Created gitea-server Custom Resource          source="gitea_reconciler.go:122"
INFO[0027] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0144] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0146] Creating VaultServer                          source="vault_reconciler.go:87"
INFO[0146] Created vault Vault Project                   source="vault_reconciler.go:94"
INFO[0146] Created vault-config Vault ConfigMap          source="vault_reconciler.go:101"
INFO[0146] Created vault Vault Service Account           source="vault_reconciler.go:109"
INFO[0147] Created vault SCC                             source="vault_reconciler.go:144"
INFO[0147] Created vault-server-binding Vault Cluster Role Binding  source="vault_reconciler.go:164"
INFO[0147] Created vault-internal Vault Service          source="vault_reconciler.go:172"
INFO[0147] Created vault Vault Service                   source="vault_reconciler.go:180"
INFO[0148] Created vault Vault Stateful                  source="vault_reconciler.go:188"
INFO[0148] Creating VaultAgent                           source="vault_reconciler.go:197"
INFO[0148] Created vault-agent-injector VaultAgent Service Account  source="vault_reconciler.go:212"
INFO[0148] Created vault-agent-injector VaultAgent Cluster Role  source="vault_reconciler.go:227"
INFO[0148] Created vault-agent-injector VaultAgent Cluster Role Binding  source="vault_reconciler.go:236"
INFO[0149] Created vault-agent-injector VaultAgent Service  source="vault_reconciler.go:245"
INFO[0149] Created vault-agent-injector VaultAgent Deployment  source="vault_reconciler.go:253"
INFO[0149] Created vault-agent-injector-cfg VaultAgent Mutating Webhook Configuration  source="vault_reconciler.go:263"
2021-10-06T18:32:33.359+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
INFO[0149] OpenShift Console URL https://console-openshift-console.apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:103"
INFO[0149] Apps Hostname Suffix apps.workshop.6hf8sw9d.kubeapp.cloud  source="workshop_controller.go:108"
INFO[0152] Waiting for deployment gitea-server. Default timeout: 420 seconds  source="k8s_helpers.go:68"
INFO[0152] Deployment 'gitea-server' successfully scaled to 1  source="k8s_helpers.go:81"
INFO[0153] Creating VaultServer                          source="vault_reconciler.go:87"
INFO[0154] Creating VaultAgent                           source="vault_reconciler.go:197"
2021-10-06T18:32:39.563+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
``` 

* delete log of vault 
```bash
INFO	controllers.Workshop	Deleting workshopcloud-native-workshop	{"workshop": "workshop-infra/cloud-native-workshop"}
INFO[0052] Deleting Vault                                source="vault_reconciler.go:271"
INFO[0052] Deleting VaultServer                          source="vault_reconciler.go:288"
INFO[0052] Deleted vault VaultServer stateful            source="vault_reconciler.go:298"
INFO[0052] Deleted vault VaultServer Service             source="vault_reconciler.go:305"
INFO[0053] Deleted vault-internal VaultServer internal Service  source="vault_reconciler.go:312"
INFO[0053] Deleted vault-server-binding  VaultServer ClusterRole Binding  source="vault_reconciler.go:320"
INFO[0053] Deleted vault vaultSCC                        source="vault_reconciler.go:327"
INFO[0053] Deleted vault VaultServer Service Account     source="vault_reconciler.go:334"
INFO[0053] Deleted vault-config VaultServer configMap    source="vault_reconciler.go:341"
INFO[0053] Deleted VaultServer Successfully              source="vault_reconciler.go:342"
INFO[0053] Deleting VaultAgent                           source="vault_reconciler.go:350"
INFO[0054] Deleted vault-agent-injector-cfg VaultAgent Mutating Webhook Configuration   source="vault_reconciler.go:363"
INFO[0054] Deleted vault-agent-injector VaultAgent Deployment   source="vault_reconciler.go:370"
INFO[0054] Deleted vault-agent-injector VaultAgent Service   source="vault_reconciler.go:378"
INFO[0054] Deleted vault-agent-injector VaultAgent Cluster Role Binding  source="vault_reconciler.go:386"
INFO[0054] Deleted vault-agent-injector VaultAgent Cluster Role  source="vault_reconciler.go:392"
INFO[0054] Deleted vault-agent-injector  VaultAgent Service Account  source="vault_reconciler.go:399"
INFO[0054] Deleted VaultAgent Successfully               source="vault_reconciler.go:400"
INFO[0054] Deleting Namespace                            source="vault_reconciler.go:408"
INFO[0055] Deleted vault Namespace                       source="vault_reconciler.go:414"
INFO[0055] Deleted Namespace Successfully                source="vault_reconciler.go:415"
INFO[0055] Deleting gitea                                source="gitea_reconciler.go:202"
INFO[0055] Deleted gitea-server gitea Custom Resource    source="gitea_reconciler.go:212"
INFO[0055] Deleted gitea-operator gitea Operator         source="gitea_reconciler.go:219"
INFO[0055] Deleted gitea-operator gitea Cluster  Role Binding  source="gitea_reconciler.go:226"
INFO[0056] Deleted gitea-operator gitea Cluster Role     source="gitea_reconciler.go:233"
INFO[0056] Deleted gitea-operator gitea Service Account  source="gitea_reconciler.go:240"
INFO[0056] Deleted giteas.gpte.opentlc.com gitea Custom Resource Definition  source="gitea_reconciler.go:247"
INFO[0056] Deleted gitea gitea Project                   source="gitea_reconciler.go:254"
INFO[0056] Gitea deleted succesfully                     source="gitea_reconciler.go:255"
INFO[0056] Deleting nexus                                source="nexus_reconciler.go:128"
I1006 18:49:50.002441  122107 request.go:621] Throttling request took 1.047380121s, request: GET:https://api.workshop.6hf8sw9d.kubeapp.cloud:6443/apis/networking.k8s.io/v1beta1?timeout=32s
INFO[0060] Finalizer removed for workshopcloud-native-workshop  source="workshop_controller.go:132"
2021-10-06T18:49:52.325+0530	DEBUG	controller	Successfully Reconciled	{"reconcilerGroup": "workshop.stakater.com", "reconcilerKind": "Workshop", "controller": "workshop", "name": "cloud-native-workshop", "namespace": "workshop-infra"}
```

### CODE REVIEW

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-40191441`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-03650b28`

File: controllers/workshop_controller.go [hunk 1]
```diff
@@ -246,10 +223,21 @@ func (r *WorkshopReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+		return result, err
+	}
+	if result, err := r.deleteVaultAgentInjector(workshop); util.IsRequeued(result, err) {
+		return result, err
+	}
+
+	if result, err := r.deleteVaultNamespace(workshop); util.IsRequeued(result, err) {
+		return result, err
+	}
+
```
Comment (rasheedamir) [inline controllers/workshop_controller.go:256]: @MaazMS this is absolutely wrong; you should add ONE method in vault which handles all these three methods

https://github.com/stakater/workshop-operator/pull/31/files#diff-239a055a1843001a63725aaf6c3ef73cc8d09de34f654bd8cd62b2d3914e2ee5R230-R240

and in the workshop controller you should just call that ONE method

File: controllers/workshop_controller.go [hunk 2]
```diff
@@ -260,3 +248,32 @@ func (r *WorkshopReconciler) handleDelete(ctx context.Context, req ctrl.Request,
 
 	return ctrl.Result{}, nil
 }
+
+func (r *WorkshopReconciler) getVariables(workshop *workshopv1.Workshop, ctx context.Context) (int, string, string, ctrl.Result, error) {
```
Comment (rasheedamir) [inline controllers/workshop_controller.go]: I don't see any benefit of it

File: controllers/workshop_controller.go (hunk 1 above)
Comment (MaazMS) [reply controllers/workshop_controller.go:256]: @rasheedamir  i will create single method, in workshop controller ,

File: controllers/workshop_controller.go (hunk 2 above)
Comment (rasheedamir) [reply controllers/workshop_controller.go]: these variables are defined only once; so, defining method for it doesn't give any value

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-4175f97d`

File: controllers/vault_reconciler.go [hunk 3]
```diff
@@ -99,72 +122,65 @@ storage "file" {
...
 	}
 
 	// Create Service
-	service := kubernetes.NewService(workshop, r.Scheme, "vault", vaultNamespace.Name, labels, []string{"http", "internal"}, []int32{8200, 8201})
+	service := kubernetes.NewService(workshop, r.Scheme, VAULTSERVICENAME, VAULTNAMESPACENAME, VaultServerlabels, []string{"http", "internal"}, []int32{8200, 8201})
 	if err := r.Create(context.TODO(), service); err != nil && !errors.IsAlreadyExists(err) {
 		return reconcile.Result{}, err
 	} else if err == nil {
-		log.Infof("Created %s Service", service.Name)
+		log.Infof("Created %s vault Service", service.Name)
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go]: lets use PascalCase i.e. `Vault` instead of `vault`

Comment (rasheedamir) [reply controllers/vault_reconciler.go]: everywhere!

File: controllers/vault_reconciler.go [hunk 4]
```diff
@@ -181,261 +197,199 @@ func (r *WorkshopReconciler) addVaultAgentInjector(workshop *workshopv1.Workshop
...
-		"app.kubernetes.io/name":    "vault",
-		"app.kubernetes.io/part-of": "vault",
-		"component":                 "server",
-	}
 
-	extraconfigFromValues := map[string]string{
-		"extraconfig-from-values.hcl": `disable_mlock = true
-ui = true
+func (r *WorkshopReconciler) deleteVault(workshop *workshopv1.Workshop) (reconcile.Result, error) {
+	log.Infoln("Deleting deleteVault ")
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go]: fix log statement

File: controllers/vault_reconciler.go [hunk 5]
```diff
@@ -181,261 +197,199 @@ func (r *WorkshopReconciler) addVaultAgentInjector(workshop *workshopv1.Workshop
...
-	configMap := kubernetes.NewConfigMap(workshop, r.Scheme, "vault-config", vaultNamespace.Name, labels, extraconfigFromValues)
-	configMapFound := &corev1.ConfigMap{}
-	configMapErr := r.Get(context.TODO(), types.NamespacedName{Name: configMap.Name, Namespace: vaultNamespace.Name}, configMapFound)
-	if configMapErr == nil {
-		// Delete configMap
-		if err := r.Delete(context.TODO(), configMap); err != nil {
-			return reconcile.Result{}, err
-		}
-		log.Infof("Deleted %s configMap", configMap.Name)
+	// Delete Service Account
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go:340]: @MaazMS we also need to remove the user from privileged SCC; as it was added here: https://github.com/stakater/workshop-operator/blob/4175f97d81e39656518a9b4c2e36a3a653cf8527/controllers/vault_reconciler.go#L114-L127

so, for proper cleanup we need to remove it from everywhere

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-1d08461e`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-e9a3304e`

File: controllers/vault_reconciler.go [hunk 6]
```diff
@@ -2,31 +2,77 @@ package controllers
...
 	"github.com/stakater/workshop-operator/common/vault"
-
-	securityv1 "github.com/openshift/api/security/v1"
-	"github.com/prometheus/common/log"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+var (
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go:17]: @MaazMS why don't we define these as constants?

File: controllers/vault_reconciler.go [hunk 7]
```diff
@@ -2,31 +2,77 @@ package controllers
...
+)
+
+const (
+	VAULT_NAMESPACE_NAME             = "vault"
+	VAULT_STATEFULSET_NAME           = "vault"
+	VAULT_SERVICE_NAME               = "vault"
+	VAULT_INTERNAL_SERVICE_NAME      = "vault-internal"
+	VAULT_ROLEBINDING_NAME           = "vault-server-binding"
+	VAULT_ROLEBINDING_ROLE_NAME      = "system:auth-delegator"
+	VAULT_ROLEBINDING_KIND_NAME      = "ClusterRole"
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go]: @MaazMS this is not vault specific! this is even more generic; so, better call it

`KIND_CLUSTER_ROLE`

File: controllers/vault_reconciler.go [hunk 8]
```diff
@@ -2,31 +2,77 @@ package controllers
...
+	VAULT_SERVICEACCOUNT_NAME        = "vault"
+	VAULT_CONFIGMAP_NAME             = "vault-config"
+	VAULTAGENT_WEBHOOK_NAME          = "vault-agent-injector-cfg"
+	VAULTAGENT_DEPLOYMENT_NAME       = "vault-agent-injector"
+	VAULTAGENT_SERVICE_NAME          = "vault-agent-injector"
+	VAULTAGENT_ROLEBINDING_NAME      = "vault-agent-injector"
+	VAULTAGENT_ROLEBINDING_KIND_NAME = "ClusterRole"
```
Comment (rasheedamir) [inline controllers/vault_reconciler.go]: this will be duplicated then!

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-839b0d56`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-8b67cead`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-1758b39a`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-c2bb3ce1`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-6d783e6e`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-232b6e64`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-ebd944bf`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-90ee2b16`

**Comment by stakater-user** [comment]
@MaazMS Image is available for testing. `docker pull stakater/workshop-operator:SNAPSHOT-PR-31-491947d6`

**Review by rasheedamir — APPROVED** [review]

**Pull request merged**